File: csharp/demo/demo25b-atomic.cs
/*
 * ATOMIC ACCESS
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo25B : IRunnable
{
    public void run()
    {
        ThreadStart doTask = () =>
        {
            Thread.Sleep(1000);
            Interlocked.Increment(ref Global.counter);
        };


        var lstTh = new List<Thread>();

        for (int i = 0; i < 1000; ++i)
            lstTh.Add(new Thread(doTask));


        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());


        // counter = 1000
        Console.WriteLine("counter = " + Global.counter);
    }


    class Global
    {
        public static volatile int counter = 0;
    }
}

File: cpp-std/demo25b-atomic.cpp
/*
ATOMIC ACCESS
*/


#include <iostream>
#include <vector>
#include <chrono>
#include <thread>
#include <atomic>
using namespace std;



std::atomic_int32_t counter;



void doTask() {
    std::this_thread::sleep_for(std::chrono::seconds(1));
    counter += 1;
}



int main() {
    counter = 0;

    vector<std::thread> lstTh;

    for (int i = 0; i < 1000; ++i) {
        lstTh.push_back(std::thread(doTask));
    }

    for (auto&& th : lstTh) {
        th.join();
    }

    // counter = 1000
    cout << "counter = " << counter << endl;
    return 0;
}
